Sources/Plasma/FeatureLib/pfMetalPipeline/plMetalDevice.cpp

```diff
@@ -0,0 +1,1279 @@
+{
+    if (mipmap->IsCompressed()) {
+        mipmap->SetCurrLevel(tRef->fLevels);
+        while ((mipmap->GetCurrWidth() | mipmap->GetCurrHeight()) & 0x03) {
+            tRef->fLevels--;
+            hsAssert(tRef->fLevels >= 0, "How was this ever compressed?");
+            if (tRef->fLevels < 0) {
+                tRef->fLevels = -1;
+                break;
+            }
+            mipmap->SetCurrLevel(tRef->fLevels);
+        }
+    }
+}
+
+void plMetalDevice::PopulateTexture(plMetalDevice::TextureRef* tRef, plMipmap* img, uint slice)
+{
+    if (img->IsCompressed()) {
+        /*
+         Some cubic assets have inconsistant mipmap sizes between their faces.
+         The DX pipeline maintains seperate structures noting the expected
+         mipmap sizes, and ignores the actual face sizes. This hack
+         makes the Metal pipeline ignore the actual face sizes and behave
+         as if all face sizes are equivelent to the first face. It does this
+         by computing the expected mipmap sizes on the fly.
+         This hack could be disabled if cube maps in the assets were
+         fixed to be consistant.
+         */
+#define HACK_LEVEL_SIZE 1
+
+#if HACK_LEVEL_SIZE
+        NS::UInteger width = tRef->fTexture->width();
+        NS::UInteger height = tRef->fTexture->height();
+#endif
+
+        if (tRef->fLevels == -1) {
+            hsAssert(1, "Bad texture found");
+            return;
+        }
+
+        for (int lvl = 0; lvl <= tRef->fLevels; lvl++) {
+            img->SetCurrLevel(lvl);
+#if HACK_LEVEL_SIZE
+            NS::UInteger levelWidth = (width / exp2(lvl));
+            NS::UInteger levelHeight = (height / exp2(lvl));
+#else
+            NS::UInteger levelWidth = img->GetCurrWidth();
+            NS::UInteger levelHeight = img->GetCurrHeight();
+#endif
+
+            switch (img->fDirectXInfo.fCompressionType) {
+                case plBitmap::DirectXInfo::kDXT1:
+                    tRef->fTexture->replaceRegion(MTL::Region::Make2D(0, 0, levelWidth, levelHeight), img->GetCurrLevel(), slice, img->GetCurrLevelPtr(), levelWidth * 2, 0);
+                    break;
+                case plBitmap::DirectXInfo::kDXT5:
+                    tRef->fTexture->replaceRegion(MTL::Region::Make2D(0, 0, img->GetCurrWidth(), img->GetCurrHeight()), img->GetCurrLevel(), slice, img->GetCurrLevelPtr(), img->GetCurrWidth() * 4, 0);
+                    break;
+            }
+        }
+    } else {
+        for (int lvl = 0; lvl <= tRef->fLevels; lvl++) {
+            img->SetCurrLevel(lvl);
+
+            if (img->GetCurrLevelPtr()) {
+                if (img->fUncompressedInfo.fType == plBitmap::UncompressedInfo::kRGB4444) {
+                    struct RGBA4444Component
+                    {
+                        unsigned r : 4;
+                        unsigned g : 4;
+                        unsigned b : 4;
+                        unsigned a : 4;
```

Comment:
Direct 3D supports textures with 4 bits per color component. Modern APIs don't generally supports this format - so it needs to be hand converted.

Reply:
More specifically, what is this syntax?
```C++
unsigned r : 4;
```

Reply:
It's been a bit - but I believe that's declaring it as a 4 bit type.

Reply:
Bit fields, which are an extremely ancient and cursed feature of C/C++, but _probably_ work for what we want here: https://en.cppreference.com/w/cpp/language/bit_field

Reply:
Can we use the fixed length type `uint32_t` instead?

Reply:
I took another look through in since I saw this code uses simd - and I didn't remember if simd had a built in 4 bit type. It does not. For now - I think this is preferable to some horrible bit shifting solution. I'm going to mark is as resolved - but I'm open to a better way to do this in the future.